### DISCUSSION TITLE
How to limit a function param to the ZodObject type (not ZodTypeAny)?

### DISCUSSION THREAD

**Discussion by anthonyma94** [colinhacks/zod#2023] [Q&A] [answered]
I'm trying to integrate Zod into a Table library with the following props:

```ts
type RowData = Record<string, string | boolean | number>;
type UseTableColsOpts<TData extends RowData> = {
  zodType: // what to put here?
}

function useTableCols<TData extends RowData>({ zodType }: UseTableColsOpts<TData>) {
  zodType.keyof(); // doesn't exist in z.ZodType or z.ZodTypeAny
}
```

`zodType` will always be an object of type `TData`. What type can I put in `zodType` so I can make use of the object methods?

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
type RowData = Record<string, string | boolean | number>
type UseTableColsOpts<TData extends RowData> = {
    objSchema: z.AnyZodObject
}

function useTableCols<TData extends RowData> ( { objSchema }: UseTableColsOpts<TData> ) {
    objSchema.keyof()
}
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by anthonyma94** [reply]
How can I make sure the schema follows `TData` and not just accept any zod object?

**Reply by JacobWeisenburger** [reply]
unfortunately, I don't think there is a way to do that
